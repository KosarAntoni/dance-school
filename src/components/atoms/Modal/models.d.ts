export type ModalProps = {
  isOpen?: boolean;
  ref?: RefObject<HTMLElement>;
};
export type ModalSetters = {
  open: () => void;
  close: () => void;
} & HTMLElement;
