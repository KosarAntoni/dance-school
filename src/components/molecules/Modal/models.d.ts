export type ModalProps = {
  justify?: 'left' | 'center' | 'right' | string;
  align?: 'top' | 'center' | 'bottom' | string;
  className?: string;
  isOpen?: boolean;
  ref?: RefObject<HTMLElement>;
};

export type ModalSetters = {
  open: () => void;
  close: () => void;
} & HTMLElement;
