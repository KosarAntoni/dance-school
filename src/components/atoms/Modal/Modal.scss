@import 'styles/component-set.scss';

@keyframes fade-in {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

@keyframes fade-out {
  0% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
}

@keyframes slide-in {
  0% {
    transform: translateY(150%) rotate(5deg);
  }

  100% {
    transform: translateY(0) rotate(0);
  }
}

@keyframes slide-out {
  0% {
    transform: translateY(0) rotate(0);
  }

  100% {
    transform: translateY(-150%) rotate(-5deg);
  }
}

.modal {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: map-get($z-index, 'modalOverlay');
  width: 100%;
  padding-top: 20px;
  padding-bottom: 20px;
  background-color: rgba(0, 0, 0, 0.2);
  animation: fade-in $transition 1 forwards;
  backdrop-filter: blur(15px);
  @include center-content();

  &__content {
    position: relative;
    z-index: map-get($z-index, 'modal');
    height: 100%;
    padding: 20px;
    border-radius: 8px;
    transform: translateY(150%) rotate(5deg);
    animation: slide-in $transition * 1.5 $transition 1 cubic-bezier(0.39, 0.8, 0.28, 1.07) forwards;
  }

  &__close {
    position: absolute;
    top: 20px;
    right: 20px;
    z-index: inherit;
  }

  &--fade-out {
    animation: fade-out $transition $transition 1 linear forwards;

    .modal {
      &__content {
        animation: slide-out $transition * 1.5 ease-out forwards;
      }
    }
  }
}
