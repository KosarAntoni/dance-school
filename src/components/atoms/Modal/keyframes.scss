@keyframes fade-in {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

@keyframes fade-out {
  0% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
}

@keyframes slide-in {
  0% {
    transform: translateY(100vh);
  }

  100% {
    transform: translateY(0);
  }
}

@keyframes slide-out {
  0% {
    transform: translateY(0);
  }

  100% {
    transform: translateY(-100vh);
  }
}

@keyframes slide-in-left {
  0% {
    transform: translateX(-150%);
  }

  100% {
    transform: translateX(0);
  }
}

@keyframes slide-out-left {
  0% {
    transform: translateX(0);
  }

  100% {
    transform: translateX(-150%);
  }
}

@keyframes slide-in-right {
  0% {
    transform: translateX(150%);
  }

  100% {
    transform: translateX(0);
  }
}

@keyframes slide-out-right {
  0% {
    transform: translateX(0);
  }

  100% {
    transform: translateX(150%);
  }
}

@keyframes slide-in-top {
  0% {
    transform: translateY(-150%);
  }

  100% {
    transform: translateX(0);
  }
}

@keyframes slide-out-top {
  0% {
    transform: translateX(0);
  }

  100% {
    transform: translateY(-150%);
  }
}

@keyframes slide-in-bottom {
  0% {
    transform: translateY(150%);
  }

  100% {
    transform: translateX(0);
  }
}

@keyframes slide-out-bottom {
  0% {
    transform: translateX(0);
  }

  100% {
    transform: translateY(150%);
  }
}
