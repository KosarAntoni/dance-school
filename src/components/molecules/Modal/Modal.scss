@use 'sass:map';
@import 'styles/component-set';
@import './keyframes';

.modal {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: map.get($z-index, 'modaloverlay');
  width: 100%;
  padding-top: 20px;
  padding-bottom: 20px;
  background-color: rgb(0 0 0 / 20%);
  animation: fade-in $transition 1 forwards;
  backdrop-filter: blur(15px);
  @include flex-center;

  &__inner {
    position: relative;
    z-index: map.get($z-index, 'modal');
    padding: 20px;
    border-radius: 8px;
  }

  &__close {
    position: absolute;
    top: 20px;
    right: 20px;
    z-index: inherit;
  }

  &--align {
    &-top {
      align-self: flex-start;
      transform: translateY(-150%);
      animation: slide-in-top $transition * 1.5 $transition 1 ease forwards;
    }

    &-center {
      align-self: center;
      transform: translateY(100vh);
      animation: slide-in $transition * 1.5 $transition 1 cubic-bezier(0.39, 0.8, 0.28, 1.07)
        forwards;
    }

    &-bottom {
      align-self: flex-end;
      transform: translateY(150%);
      animation: slide-in-bottom $transition * 1.5 $transition 1 ease forwards;
    }
  }

  &--justify {
    &-left {
      margin-right: auto;
      transform: translateX(-150%);
      animation: slide-in-left $transition * 1.5 $transition 1 ease forwards;
    }

    &-center {
      margin-right: auto;
      margin-left: auto;
    }

    &-right {
      margin-left: auto;
      transform: translateX(150%);
      animation: slide-in-right $transition * 1.5 $transition 1 ease forwards;
    }
  }

  &--fade-out {
    animation: fade-out $transition $transition 1 linear forwards;

    .modal {
      &--align {
        &-top {
          animation: slide-out-top $transition * 1.5 ease-out forwards;
        }

        &-center {
          animation: slide-out $transition * 1.5 ease-out forwards;
        }

        &-bottom {
          animation: slide-out-bottom $transition * 1.5 ease-out forwards;
        }
      }

      &--justify {
        &-left {
          animation: slide-out-left $transition * 1.5 ease-out forwards;
        }

        &-right {
          animation: slide-out-right $transition * 1.5 ease-out forwards;
        }
      }
    }
  }
}
