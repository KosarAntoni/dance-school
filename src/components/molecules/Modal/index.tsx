export { default } from './Modal';
export * from './models.d';
