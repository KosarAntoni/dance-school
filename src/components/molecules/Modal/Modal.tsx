import React, {
  FC,
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';
import { createPortal } from 'react-dom';
import { HiX } from '@react-icons/all-files/hi/HiX';
import classNames from 'classnames';

import Button from 'atoms/Button';
import Container from 'atoms/Container';

import { isBrowser } from 'utils/browser';

import { ModalProps } from './models';

import './Modal.scss';

const Modal: FC<ModalProps> = forwardRef(
  ({ children, justify = 'center', align = 'center', isOpen, className }, ref) => {
    const bodyElement = isBrowser() ? window.document.body : null;
    const modalElement = isBrowser() ? document.getElementById('modal-root') : null;
    const [isModalOpen, setIsModalOpen] = useState<boolean | undefined>(isOpen);
    const [isFadeOut, setIsFadeOut] = useState<boolean | undefined>(false);
    const contentRef = useRef<HTMLDivElement>(null);

    const modalOuterClasses = classNames('modal', { 'modal--fade-out': isFadeOut });
    const modalClasses = classNames(
      'modal__inner',
      'background--white',
      { [`modal--align-${align}`]: align, [`modal--justify-${justify}`]: justify },
      className
    );

    const close = useCallback(() => {
      setIsFadeOut(true);
      setTimeout(() => {
        if (bodyElement) bodyElement.classList.remove('scroll--prevent');
        setIsModalOpen(false);
        setIsFadeOut(false);
      }, 600);
    }, [bodyElement]);

    const open = useCallback(() => {
      if (bodyElement) bodyElement.classList.add('scroll--prevent');
      setIsModalOpen(true);
    }, [bodyElement]);

    useImperativeHandle(
      ref,
      () => ({
        open,
        close,
      }),
      [close, open]
    );

    const onEscPress = useCallback(
      (event) => {
        if (event.keyCode === 27) close();
      },
      [close]
    );

    const onClickOutside = useCallback(
      (event) => {
        if (contentRef.current && !contentRef.current.contains(event.target)) {
          close();
        }
      },
      [close]
    );

    useEffect(() => {
      if (isModalOpen) {
        document.addEventListener('keydown', onEscPress, false);
        document.addEventListener('click', onClickOutside, true);
      }

      return () => {
        document.removeEventListener('keydown', onEscPress, false);
        document.removeEventListener('click', onClickOutside, true);
      };
    }, [onEscPress, isModalOpen, isOpen, onClickOutside]);

    if (!modalElement) return null;

    const renderModal = isModalOpen ? (
      <Container
        variant="fluid"
        innerRef={contentRef}
        outerClassName={modalOuterClasses}
        className={modalClasses}
        data-testid="modal-item"
      >
        <Button
          onClick={close}
          className="modal__close"
          variant="solid"
          color="black"
          size={18}
          padding="xs"
        >
          <HiX />
        </Button>
        {children}
      </Container>
    ) : null;

    return createPortal(renderModal, modalElement);
  }
);

export default Modal;
