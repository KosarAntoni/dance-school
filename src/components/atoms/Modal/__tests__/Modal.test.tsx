import React from 'react';
import { act } from 'react-dom/test-utils';
import { fireEvent, render, screen } from '@testing-library/react';

import { ModalSetters } from '..';
import Modal from '../Modal';

jest.useFakeTimers();
jest.spyOn(global, 'setTimeout');

const testModalCLose = (node: HTMLElement) => {
  // change class after click on background
  expect(node).toHaveClass('modal--fade-out');

  // after 0.45 second element is removed from DOM
  expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 600);
  act(() => {
    jest.advanceTimersByTime(600);
  });
  expect(node).not.toBeInTheDocument();
};

test('the Modal component renders correctly with required and optional props', () => {
  // add a div with #modal-root id to the global body
  const modalRoot = global.document.createElement('div');
  modalRoot.setAttribute('id', 'modal-root');
  const body = global.document.querySelector('body');
  if (body) body.appendChild(modalRoot);

  // create element ref to test forwardRef hook
  const modalRef = React.createRef<ModalSetters>();

  render(
    <Modal ref={modalRef} isOpen>
      TEST TEXT
    </Modal>
  );

  let modalNode = screen.getByTestId('modal-item');

  // the component has been rendered
  expect(modalNode).toBeInTheDocument();

  // generating a snapshot, not for testing purposes but only to see what changes in the HTML structure during CR if the developer updates the snapshot
  expect(modalNode).toMatchSnapshot();

  // the component has content inside
  expect(screen.getByText('TEST TEXT')).toBeInTheDocument();

  // modal closes after click on background
  fireEvent.click(modalNode);
  testModalCLose(modalNode);

  // after ref call element is inside DOM
  act(() => {
    if (modalRef.current) modalRef.current.open();
  });

  modalNode = screen.getByTestId('modal-item');
  expect(modalNode).toBeInTheDocument();

  // modal closes after click on close button
  const closeButtonNode = screen.getByRole('button');
  fireEvent.click(closeButtonNode);

  testModalCLose(modalNode);

  // open modal for testing purposes
  act(() => {
    if (modalRef.current) modalRef.current.open();
  });

  // modal closes after press on Escape key
  fireEvent.keyDown(modalRoot, { key: 'Escape', keyCode: 27 });
  testModalCLose(modalNode);
});
