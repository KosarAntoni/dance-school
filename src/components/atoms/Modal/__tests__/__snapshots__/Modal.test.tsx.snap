// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Modal component renders correctly with required and optional props 1`] = `
<div
  class="container modal__inner background--white modal--align-center modal--justify-center"
  data-testid="modal-item"
>
  <button
    class="button modal__close button--solid-black button--around"
    type="button"
  >
    <span
      class="typography button__content typography--size-18 typography--padding-xs typography--line-height-normal"
      data-testid="typography-item"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </button>
  TEST TEXT
</div>
`;
